File: CPP/vendingmachine/src/ReturnChangeState.cpp

#include "ReturnChangeState.h"
#include "VendingMachine.h"
#include <iostream>

void ReturnChangeState::selectProduct(VendingMachine *vendingMachine,
                                      const Product &product)
{
  std::cout << "Please collect the change first." << std::endl;
}

void ReturnChangeState::insertCoin(VendingMachine *vendingMachine, Coin coin)
{
  std::cout << "Please collect the change first." << std::endl;
}

void ReturnChangeState::insertNote(VendingMachine *vendingMachine, Note note)
{
  std::cout << "Please collect the change first." << std::endl;
}

void ReturnChangeState::dispenseProduct(VendingMachine *vendingMachine)
{
  std::cout << "Product already dispensed. Please collect the change."
            << std::endl;
}

void ReturnChangeState::returnChange(VendingMachine *vendingMachine)
{
  double change = vendingMachine->totalPayment -
                  vendingMachine->selectedProduct->getPrice();
  if (change > 0)
  {
    std::cout << "Change returned: " << change << "Rs." << std::endl;
    vendingMachine->resetPayment();
  }
  else
  {
    std::cout << "No change to return." << std::endl;
  }
  vendingMachine->resetSelectedProduct();
  vendingMachine->setState(&vendingMachine->idleState);
}

File: JAVA/vendingmachine/ReturnChangeState.java
package vendingmachine;

public class ReturnChangeState implements VendingMachineState {
    private final VendingMachine vendingMachine;
    public ReturnChangeState(VendingMachine vendingMachine) {
        this.vendingMachine = vendingMachine;
    }

    @Override
    public void selectProduct(Product product) {
        System.out.println("Product already selected: " + vendingMachine.getSelectedProduct().getName() + ". Please collect your cchange frist.");
    }

    @Override
    public void insertCoin(Coin coin) {
        System.out.println("Payment already made. Please collect your change.");
    }

    @Override
    public void insertNote(Note note) {
        System.out.println("Payment already made. Please collect your change.");
    }

    @Override
    public void dispenseProduct() {
        System.out.println("Product is already dispensed. Please collect your change.");
    }

    @Override
    public void returnChange() {
        int change = vendingMachine.getTotalPayment() - vendingMachine.getSelectedProduct().getPrice();
        if(change > 0) {
            System.out.println("Change returned: " + change);
            vendingMachine.resetPayment();
        }
        else{
            System.out.println("No change to return.");
        }
        vendingMachine.resetSelectedProduct();
        vendingMachine.setState(vendingMachine.getIdleState());
    }  
}
